File: r/R/portal.R
#' Leitura dos artefatos de emprego publicados no GitHub Pages
#'
#' Lado leitor do pipeline de emprego. O projeto `cgid_cargas` busca CAGED/RAIS
#' (via a API desta biblioteca), gera dois artefatos por estado e os publica no
#' GitHub Pages, em dois ambientes (`producao` e `homologacao`):
#'
#' \preformatted{
#'   Portal    : <BASE>/{ambiente}/{UF}/emprego.json             (schema v1.1)
#'   Relatorio : <BASE>/{ambiente}/cache/{UF}/emprego_cache.json (schema 1.0)
#' }
#'
#' Estas funcoes leem esses JSONs de volta e os entregam como `tibble`/`list`:
#'   - Relatorio (cache 1.0): dados brutos `saldo` (CAGED mensal por divisao
#'     CNAE), `estoque` (RAIS por divisao) e `acum12m_total`. Para analise.
#'   - Portal (v1.1): payload pronto para apresentacao (`capa_kpi`, `kpis`,
#'     `charts`, `ranked_lists`, `breakdowns`), ja em pt-BR.
#'
#' Configuracao via variaveis de ambiente:
#'   `PORTAL_EMPREGO_BASE_URL` (URL base) e `PORTAL_EMPREGO_AMBIENTE` (ambiente).
#'
#' @importFrom httr2 request req_timeout req_headers req_error req_perform resp_status resp_body_json
#' @importFrom tibble as_tibble tibble
#' @importFrom rlang abort
#' @name portal_emprego
NULL

# URL publica base onde o cgid_cargas publica os artefatos de emprego.
.PORTAL_EMPREGO_BASE_URL_DEFAULT <- "https://amiltonmendes.github.io/sdic_libraries"
.PORTAL_EMPREGO_UFS_27 <- c(
  "AC", "AL", "AP", "AM", "BA", "CE", "DF", "ES", "GO", "MA", "MT", "MS",
  "MG", "PA", "PB", "PR", "PE", "PI", "RJ", "RN", "RS", "RO", "RR", "SC",
  "SP", "SE", "TO"
)

# ── Helpers internos ──────────────────────────────────────────────────────────

.portal_emprego_base_url <- function() {
  url <- Sys.getenv("PORTAL_EMPREGO_BASE_URL", unset = "")
  if (!nzchar(url)) url <- .PORTAL_EMPREGO_BASE_URL_DEFAULT
  sub("/+$", "", url)
}

.portal_emprego_validar_ambiente <- function(ambiente) {
  amb <- tolower(trimws(ambiente %||% "producao"))
  if (!amb %in% c("producao", "homologacao")) {
    rlang::abort(sprintf(
      "ambiente invalido: '%s'. Use 'producao' ou 'homologacao'.", ambiente
    ))
  }
  amb
}

.portal_emprego_path_portal <- function(uf, ambiente) {
  sprintf("%s/%s/emprego.json", ambiente, toupper(uf))
}

.portal_emprego_path_relatorio <- function(uf, ambiente) {
  sprintf("%s/cache/%s/emprego_cache.json", ambiente, toupper(uf))
}

# GET de um JSON publicado. 404 -> NULL se opcional, senao erro.
.portal_emprego_fetch_json <- function(path, opcional = FALSE) {
  timeout <- as.integer(Sys.getenv("API_TIMEOUT", unset = "30"))
  version <- Sys.getenv("SDIC_VERSION", unset = "0.4.0")
  url <- paste0(.portal_emprego_base_url(), "/", sub("^/+", "", path))

  resp <- tryCatch(
    httr2::request(url) |>
      httr2::req_timeout(timeout) |>
      httr2::req_headers(
        Accept = "application/json",
        `User-Agent` = paste0("sdic-libraries/", version)
      ) |>
      httr2::req_error(is_error = function(resp) FALSE) |>
      httr2::req_perform(),
    error = function(e) {
      if (opcional) return(NULL)
      rlang::abort(paste0(
        "Problema de conectividade ao buscar dados publicados. ",
        "Verifique sua conexao e tente novamente."
      ))
    }
  )
  if (is.null(resp)) return(NULL)

  status <- httr2::resp_status(resp)
  if (status == 404) {
    if (opcional) return(NULL)
    rlang::abort(sprintf("Artefato nao encontrado (404): %s", path))
  }
  if (status >= 400) {
    rlang::abort(sprintf("Falha ao buscar artefato publicado (HTTP %s).", status))
  }
  httr2::resp_body_json(resp, simplifyVector = TRUE)
}

# Reforca dtypes do saldo (competencia/cnae_divisao como character com zero a esquerda).
.portal_emprego_df_saldo <- function(payload) {
  saldo <- if (is.null(payload)) NULL else payload$saldo
  if (is.null(saldo) || length(saldo) == 0) return(tibble::tibble())
  df <- tibble::as_tibble(saldo)
  if ("competencia" %in% names(df)) {
    df$competencia <- formatC(as.integer(df$competencia), width = 6, format = "d", flag = "0")
  }
  if ("cnae_divisao" %in% names(df)) {
    df$cnae_divisao <- formatC(as.integer(df$cnae_divisao), width = 2, format = "d", flag = "0")
  }
  df
}

.portal_emprego_df_estoque <- function(payload) {
  estoque <- if (is.null(payload)) NULL else payload$estoque
  if (is.null(estoque) || length(estoque) == 0) return(tibble::tibble())
  df <- tibble::as_tibble(estoque)
  if ("cnae_divisao" %in% names(df)) {
    df$cnae_divisao <- formatC(as.integer(df$cnae_divisao), width = 2, format = "d", flag = "0")
  }
  df
}

# Lista de objetos (ranked_lists/breakdowns) -> lista nomeada de tibbles.
# `simplifyVector` entrega um data.frame com `items` como list-column de data.frames.
.portal_emprego_named_tibbles <- function(itens, campo_itens, prefixo) {
  out <- list()
  if (is.null(itens) || length(itens) == 0) return(out)
  if (is.data.frame(itens)) {
    n <- nrow(itens)
    ids <- if (!is.null(itens$id)) as.character(itens$id) else rep(NA_character_, n)
    conteudos <- itens[[campo_itens]]
    for (i in seq_len(n)) {
      id <- if (!is.na(ids[i]) && nzchar(ids[i])) ids[i] else sprintf("%s_%d", prefixo, i)
      conteudo <- if (is.list(conteudos)) conteudos[[i]] else conteudos
      out[[id]] <- if (is.null(conteudo) || length(conteudo) == 0) tibble::tibble() else tibble::as_tibble(conteudo)
    }
  } else {
    for (i in seq_along(itens)) {
      el <- itens[[i]]
      id <- el$id %||% sprintf("%s_%d", prefixo, i)
      conteudo <- el[[campo_itens]]
      out[[id]] <- if (is.null(conteudo) || length(conteudo) == 0) tibble::tibble() else tibble::as_tibble(conteudo)
    }
  }
  out
}

# ── RELATORIO (cache 1.0 — dados brutos) ──────────────────────────────────────

#' Relatorio completo de emprego de uma UF (cache 1.0)
#'
#' @param uf Sigla da UF (ex.: "SP").
#' @param ambiente "producao" (default) ou "homologacao".
#' @return `list(saldo, estoque, acum12m_total, metadata)`.
#' @export
get_relatorio_emprego_estadual <- function(uf, ambiente = "producao") {
  amb <- .portal_emprego_validar_ambiente(ambiente)
  payload <- .portal_emprego_fetch_json(.portal_emprego_path_relatorio(uf, amb))
  acum <- if (is.null(payload$acum12m_total)) NULL else as.integer(payload$acum12m_total)
  list(
    saldo = .portal_emprego_df_saldo(payload),
    estoque = .portal_emprego_df_estoque(payload),
    acum12m_total = acum,
    metadata = payload$metadata %||% list()
  )
}

#' Saldo CAGED mensal por divisao CNAE (cache) de uma UF
#' @inheritParams get_relatorio_emprego_estadual
#' @return Tibble com o saldo mensal.
#' @export
get_relatorio_emprego_saldo_estadual <- function(uf, ambiente = "producao") {
  amb <- .portal_emprego_validar_ambiente(ambiente)
  payload <- .portal_emprego_fetch_json(.portal_emprego_path_relatorio(uf, amb))
  .portal_emprego_df_saldo(payload)
}

#' Estoque RAIS por divisao CNAE (cache) de uma UF
#' @inheritParams get_relatorio_emprego_estadual
#' @return Tibble com o estoque.
#' @export
get_relatorio_emprego_estoque_estadual <- function(uf, ambiente = "producao") {
  amb <- .portal_emprego_validar_ambiente(ambiente)
  payload <- .portal_emprego_fetch_json(.portal_emprego_path_relatorio(uf, amb))
  .portal_emprego_df_estoque(payload)
}

#' Metadados do relatorio (reference, fonte, gerado_em, ...) de uma UF
#' @inheritParams get_relatorio_emprego_estadual
#' @return Lista de metadados.
#' @export
get_relatorio_emprego_metadata_estadual <- function(uf, ambiente = "producao") {
  amb <- .portal_emprego_validar_ambiente(ambiente)
  payload <- .portal_emprego_fetch_json(.portal_emprego_path_relatorio(uf, amb))
  payload$metadata %||% list()
}

#' Saldo de todas as UFs disponiveis, empilhado, com coluna `uf`
#' @param ambiente "producao" (default) ou "homologacao".
#' @return Tibble empilhado.
#' @export
get_relatorio_emprego_saldo_todos_estados <- function(ambiente = "producao") {
  amb <- .portal_emprego_validar_ambiente(ambiente)
  frames <- list()
  for (uf in .PORTAL_EMPREGO_UFS_27) {
    payload <- .portal_emprego_fetch_json(.portal_emprego_path_relatorio(uf, amb), opcional = TRUE)
    df <- .portal_emprego_df_saldo(payload)
    if (nrow(df) > 0) {
      df <- cbind(uf = uf, df)
      frames[[length(frames) + 1]] <- df
    }
  }
  if (length(frames) == 0) return(tibble::tibble())
  tibble::as_tibble(dplyr::bind_rows(frames))
}

#' Estoque de todas as UFs disponiveis, empilhado, com coluna `uf`
#' @param ambiente "producao" (default) ou "homologacao".
#' @return Tibble empilhado.
#' @export
get_relatorio_emprego_estoque_todos_estados <- function(ambiente = "producao") {
  amb <- .portal_emprego_validar_ambiente(ambiente)
  frames <- list()
  for (uf in .PORTAL_EMPREGO_UFS_27) {
    payload <- .portal_emprego_fetch_json(.portal_emprego_path_relatorio(uf, amb), opcional = TRUE)
    df <- .portal_emprego_df_estoque(payload)
    if (nrow(df) > 0) {
      df <- cbind(uf = uf, df)
      frames[[length(frames) + 1]] <- df
    }
  }
  if (length(frames) == 0) return(tibble::tibble())
  tibble::as_tibble(dplyr::bind_rows(frames))
}

# ── PORTAL (payload v1.1) ─────────────────────────────────────────────────────

#' Payload completo do portal (v1.1) de uma UF
#' @inheritParams get_relatorio_emprego_estadual
#' @return Lista com o payload do portal, como publicado.
#' @export
get_portal_emprego_estadual <- function(uf, ambiente = "producao") {
  amb <- .portal_emprego_validar_ambiente(ambiente)
  .portal_emprego_fetch_json(.portal_emprego_path_portal(uf, amb)) %||% list()
}

#' KPI de capa (`capa_kpi`) de uma UF
#' @inheritParams get_relatorio_emprego_estadual
#' @return Lista com o KPI de capa.
#' @export
get_portal_emprego_capa_estadual <- function(uf, ambiente = "producao") {
  payload <- get_portal_emprego_estadual(uf, ambiente)
  payload$capa_kpi %||% list()
}

#' KPIs do portal de uma UF como tibble
#' @inheritParams get_relatorio_emprego_estadual
#' @return Tibble de KPIs.
#' @export
get_portal_emprego_kpis_estadual <- function(uf, ambiente = "producao") {
  payload <- get_portal_emprego_estadual(uf, ambiente)
  if (is.null(payload$kpis) || length(payload$kpis) == 0) return(tibble::tibble())
  tibble::as_tibble(payload$kpis)
}

#' Series dos graficos do portal de uma UF
#' @inheritParams get_relatorio_emprego_estadual
#' @return Lista nomeada `{chart_id: tibble(competencia, valor)}`.
#' @export
get_portal_emprego_charts_estadual <- function(uf, ambiente = "producao") {
  payload <- get_portal_emprego_estadual(uf, ambiente)
  charts <- payload$charts
  out <- list()
  if (is.null(charts) || length(charts) == 0) return(out)

  # `series` é list-column; cada elemento é um data.frame de séries, cuja
  # coluna `points` é, por sua vez, um data.frame com colunas t/v.
  primeira_serie_points <- function(serie) {
    if (is.null(serie)) return(NULL)
    if (is.data.frame(serie)) {
      pts <- serie$points
      if (is.list(pts)) pts[[1]] else pts
    } else {
      s1 <- serie[[1]]
      s1$points
    }
  }
  monta_pts <- function(pts) {
    if (is.null(pts) || length(pts) == 0) return(tibble::tibble())
    p <- tibble::as_tibble(pts)
    tibble::tibble(competencia = p$t, valor = p$v)
  }

  if (is.data.frame(charts)) {
    n <- nrow(charts)
    ids <- if (!is.null(charts$id)) as.character(charts$id) else rep(NA_character_, n)
    series_col <- charts$series
    for (i in seq_len(n)) {
      id <- if (!is.na(ids[i]) && nzchar(ids[i])) ids[i] else sprintf("chart_%d", i)
      serie <- if (is.list(series_col)) series_col[[i]] else series_col
      out[[id]] <- monta_pts(primeira_serie_points(serie))
    }
  } else {
    for (i in seq_along(charts)) {
      ch <- charts[[i]]
      id <- ch$id %||% sprintf("chart_%d", i)
      out[[id]] <- monta_pts(primeira_serie_points(ch$series))
    }
  }
  out
}

#' Rankings do portal (maiores saldos/estoques) de uma UF
#' @inheritParams get_relatorio_emprego_estadual
#' @return Lista nomeada `{list_id: tibble(items)}`.
#' @export
get_portal_emprego_ranked_lists_estadual <- function(uf, ambiente = "producao") {
  payload <- get_portal_emprego_estadual(uf, ambiente)
  .portal_emprego_named_tibbles(payload$ranked_lists, "items", "rank")
}

#' Composicoes do portal (ex.: intensidade tecnologica) de uma UF
#' @inheritParams get_relatorio_emprego_estadual
#' @return Lista nomeada `{id: tibble(items)}`.
#' @export
get_portal_emprego_breakdowns_estadual <- function(uf, ambiente = "producao") {
  payload <- get_portal_emprego_estadual(uf, ambiente)
  .portal_emprego_named_tibbles(payload$breakdowns, "items", "breakdown")
}

#' KPIs de todas as UFs disponiveis, empilhados, com coluna `uf`
#' @param ambiente "producao" (default) ou "homologacao".
#' @return Tibble empilhado — comparacao entre estados.
#' @export
get_portal_emprego_kpis_todos_estados <- function(ambiente = "producao") {
  amb <- .portal_emprego_validar_ambiente(ambiente)
  frames <- list()
  for (uf in .PORTAL_EMPREGO_UFS_27) {
    payload <- .portal_emprego_fetch_json(.portal_emprego_path_portal(uf, amb), opcional = TRUE)
    if (!is.null(payload$kpis) && length(payload$kpis) > 0) {
      df <- cbind(uf = uf, tibble::as_tibble(payload$kpis))
      frames[[length(frames) + 1]] <- df
    }
  }
  if (length(frames) == 0) return(tibble::tibble())
  tibble::as_tibble(dplyr::bind_rows(frames))
}

# ── Descoberta ────────────────────────────────────────────────────────────────

#' Lista as UFs com artefato publicado
#'
#' @param ambiente "producao" (default) ou "homologacao".
#' @param tipo "portal" (emprego.json) ou "relatorio" (cache).
#' @return Vetor de siglas de UF disponiveis.
#' @export
listar_estados_disponiveis <- function(ambiente = "producao", tipo = "portal") {
  amb <- .portal_emprego_validar_ambiente(ambiente)
  disponiveis <- character(0)
  for (uf in .PORTAL_EMPREGO_UFS_27) {
    path <- if (identical(tipo, "relatorio")) {
      .portal_emprego_path_relatorio(uf, amb)
    } else {
      .portal_emprego_path_portal(uf, amb)
    }
    payload <- .portal_emprego_fetch_json(path, opcional = TRUE)
    if (!is.null(payload)) disponiveis <- c(disponiveis, uf)
  }
  disponiveis
}

File: python/sdic_libraries/dados/emprego/portal.py
"""portal.py — leitura dos artefatos de emprego publicados no GitHub Pages.

Lado **leitor** do pipeline de emprego. O projeto `cgid_cargas` busca CAGED/RAIS
(via a API desta biblioteca), gera dois artefatos por estado e os publica no
GitHub Pages, em dois ambientes (`producao` e `homologacao`):

    Portal    : <BASE>/{ambiente}/{UF}/emprego.json            (schema v1.1)
    Relatório : <BASE>/{ambiente}/cache/{UF}/emprego_cache.json (schema 1.0)

Este módulo lê esses JSONs de volta e os entrega como `dict`/`pandas.DataFrame`:

  - **Relatório** (cache 1.0): dados brutos — `saldo` (CAGED mensal por divisão
    CNAE), `estoque` (RAIS por divisão) e `acum12m_total`. Para análise.
  - **Portal** (v1.1): payload pronto para apresentação — `capa_kpi`, `story`,
    `kpis`, `charts`, `ranked_lists`, `breakdowns`, já em pt-BR.

Configuração:
    PORTAL_EMPREGO_BASE_URL   URL pública base (default abaixo).
    PORTAL_EMPREGO_AMBIENTE   ambiente default ('producao' | 'homologacao').

Uso:
    from sdic_libraries.dados.emprego.portal import (
        get_relatorio_emprego_saldo_estadual,
        get_portal_emprego_estadual,
    )
    df = get_relatorio_emprego_saldo_estadual("SP")
    payload = get_portal_emprego_estadual("SP", ambiente="homologacao")
"""
from __future__ import annotations

import logging
import os
from pathlib import Path
from typing import Any, Dict, List, Optional

import pandas as pd
import requests

# Reaproveita a exceção e a tradução amigável de erros do cliente da API.
from .api import EmpregoAPIError, _get_user_friendly_error_message

# URL pública base onde o `cgid_cargas` publica os artefatos de emprego.
DEFAULT_PORTAL_BASE_URL = "https://amiltonmendes.github.io/sdic_libraries"
AMBIENTES = ("producao", "homologacao")

UFS_27 = [
    "AC", "AL", "AP", "AM", "BA", "CE", "DF", "ES", "GO", "MA", "MT", "MS",
    "MG", "PA", "PB", "PR", "PE", "PI", "RJ", "RN", "RS", "RO", "RR", "SC",
    "SP", "SE", "TO",
]


class PortalEmprego:
    """Cliente de leitura dos artefatos de emprego publicados no GitHub Pages.

    Attributes:
        base_url (str): URL pública base (sem `/{ambiente}/...`).
        ambiente (str): 'producao' (default) ou 'homologacao'.
        timeout (int): timeout das requisições em segundos.
        session (requests.Session): sessão HTTP com pool de conexões.
    """

    def __init__(
        self,
        base_url: Optional[str] = None,
        ambiente: Optional[str] = None,
        timeout: int = 30,
    ):
        self._load_env_files()

        self.base_url = (
            base_url
            or os.getenv("PORTAL_EMPREGO_BASE_URL")
            or DEFAULT_PORTAL_BASE_URL
        ).rstrip("/")

        self.ambiente = self._validar_ambiente(
            ambiente or os.getenv("PORTAL_EMPREGO_AMBIENTE") or "producao"
        )
        self.timeout = int(os.getenv("API_TIMEOUT", str(timeout)))

        log_level = os.getenv("LOG_LEVEL", "INFO").upper()
        if hasattr(logging, log_level):
            logging.getLogger().setLevel(getattr(logging, log_level))

        self.session = requests.Session()
        self.logger = logging.getLogger(__name__)
        version = os.getenv("SDIC_VERSION", "0.4.0")
        self.session.headers.update({
            "User-Agent": f"sdic-libraries/{version}",
            "Accept": "application/json",
        })

    # ── Contexto / env ────────────────────────────────────────────────────────

    def __enter__(self) -> "PortalEmprego":
        return self

    def __exit__(self, *exc) -> None:
        self.session.close()

    @staticmethod
    def _validar_ambiente(ambiente: str) -> str:
        amb = (ambiente or "").strip().lower()
        if amb not in AMBIENTES:
            raise ValueError(
                f"ambiente inválido: {ambiente!r}. Use {' ou '.join(AMBIENTES)}."
            )
        return amb

    def _load_env_files(self) -> None:
        """Carrega `.env` automaticamente (mesma lógica do cliente da API)."""
        env_locations = [
            Path.cwd() / ".env",
            Path.home() / ".env",
            Path("/etc/sdic/.env"),
        ]
        try:
            from dotenv import load_dotenv

            for env_file in env_locations:
                if env_file.exists():
                    load_dotenv(env_file, override=False)
                    break
        except ImportError:
            for env_file in env_locations:
                if env_file.exists():
                    self._load_env_file(env_file)
                    break

    def _load_env_file(self, env_file_path: Path) -> None:
        try:
            with open(env_file_path, "r", encoding="utf-8") as f:
                for line in f:
                    line = line.strip()
                    if line and not line.startswith("#") and "=" in line:
                        key, value = (p.strip() for p in line.split("=", 1))
                        if value[:1] in {'"', "'"} and value[-1:] == value[:1]:
                            value = value[1:-1]
                        os.environ.setdefault(key, value)
        except Exception:  # noqa: BLE001 — experiência fluida
            pass

    # ── HTTP ──────────────────────────────────────────────────────────────────

    def _fetch_json(self, path: str, *, opcional: bool = False) -> Optional[dict]:
        """GET de um JSON publicado. 404 → None se `opcional`, senão erro amigável."""
        url = f"{self.base_url}/{path.lstrip('/')}"
        try:
            response = self.session.get(url, timeout=self.timeout)
            if opcional and response.status_code == 404:
                return None
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            status_code = getattr(getattr(e, "response", None), "status_code", None)
            if opcional and status_code == 404:
                return None
            self.logger.error(f"Falha ao buscar {path}: {e}")
            raise EmpregoAPIError(_get_user_friendly_error_message(e, status_code))
        except ValueError as e:
            self.logger.error(f"JSON inválido em {path}: {e}")
            raise EmpregoAPIError(_get_user_friendly_error_message(e))

    def _ambiente(self, ambiente: Optional[str]) -> str:
        return self._validar_ambiente(ambiente) if ambiente else self.ambiente

    # ── Caminhos ──────────────────────────────────────────────────────────────

    def _path_portal(self, uf: str, ambiente: str) -> str:
        return f"{ambiente}/{uf.upper()}/emprego.json"

    def _path_relatorio(self, uf: str, ambiente: str) -> str:
        return f"{ambiente}/cache/{uf.upper()}/emprego_cache.json"

    # ── Relatório (cache 1.0 — dados brutos) ──────────────────────────────────

    def relatorio(
        self, uf: str, ambiente: Optional[str] = None, *, opcional: bool = False
    ) -> Optional[dict]:
        """Payload bruto do cache (`emprego_cache.json`) da UF. None se ausente e opcional."""
        amb = self._ambiente(ambiente)
        return self._fetch_json(self._path_relatorio(uf, amb), opcional=opcional)

    # ── Portal (v1.1 — apresentação) ──────────────────────────────────────────

    def portal(
        self, uf: str, ambiente: Optional[str] = None, *, opcional: bool = False
    ) -> Optional[dict]:
        """Payload do portal (`emprego.json`, v1.1) da UF. None se ausente e opcional."""
        amb = self._ambiente(ambiente)
        return self._fetch_json(self._path_portal(uf, amb), opcional=opcional)


# ── Reconstrução de DataFrames a partir do cache (schema 1.0) ──────────────────

def _df_saldo(payload: Optional[dict]) -> pd.DataFrame:
    """`saldo` → DataFrame, com dtypes reforçados (competencia/cnae_divisao string)."""
    df = pd.DataFrame((payload or {}).get("saldo") or [])
    if not df.empty:
        if "competencia" in df.columns:
            df["competencia"] = df["competencia"].astype(str).str.zfill(6)
        if "cnae_divisao" in df.columns:
            df["cnae_divisao"] = df["cnae_divisao"].astype(str).str.zfill(2)
    return df


def _df_estoque(payload: Optional[dict]) -> pd.DataFrame:
    """`estoque` → DataFrame, com `cnae_divisao` como string (zero à esquerda)."""
    df = pd.DataFrame((payload or {}).get("estoque") or [])
    if not df.empty and "cnae_divisao" in df.columns:
        df["cnae_divisao"] = df["cnae_divisao"].astype(str).str.zfill(2)
    return df


def _acum12m(payload: Optional[dict]) -> Optional[int]:
    """`acum12m_total` — None quando ausente (NUNCA 0)."""
    acum = (payload or {}).get("acum12m_total", None)
    return int(acum) if acum is not None else None


# ── Funções de módulo: RELATÓRIO (dados brutos do cache) ───────────────────────

def get_relatorio_emprego_estadual(uf: str, ambiente: str = "producao") -> dict:
    """Relatório completo (cache 1.0) de uma UF.

    Args:
        uf (str): sigla da UF (ex.: 'SP').
        ambiente (str): 'producao' (default) ou 'homologacao'.

    Returns:
        dict: {'saldo': DataFrame, 'estoque': DataFrame,
               'acum12m_total': int | None, 'metadata': dict}
    """
    with PortalEmprego(ambiente=ambiente) as p:
        payload = p.relatorio(uf) or {}
    return {
        "saldo": _df_saldo(payload),
        "estoque": _df_estoque(payload),
        "acum12m_total": _acum12m(payload),
        "metadata": payload.get("metadata", {}),
    }


def get_relatorio_emprego_saldo_estadual(uf: str, ambiente: str = "producao") -> pd.DataFrame:
    """Saldo CAGED mensal por divisão CNAE (cache) de uma UF, como DataFrame."""
    with PortalEmprego(ambiente=ambiente) as p:
        return _df_saldo(p.relatorio(uf))


def get_relatorio_emprego_estoque_estadual(uf: str, ambiente: str = "producao") -> pd.DataFrame:
    """Estoque RAIS por divisão CNAE (cache) de uma UF, como DataFrame."""
    with PortalEmprego(ambiente=ambiente) as p:
        return _df_estoque(p.relatorio(uf))


def get_relatorio_emprego_metadata_estadual(uf: str, ambiente: str = "producao") -> dict:
    """Metadados do relatório (reference, fonte, gerado_em, ...) de uma UF."""
    with PortalEmprego(ambiente=ambiente) as p:
        return (p.relatorio(uf) or {}).get("metadata", {})


def get_relatorio_emprego_saldo_todos_estados(ambiente: str = "producao") -> pd.DataFrame:
    """Saldo de todas as UFs disponíveis, empilhado, com coluna `uf`."""
    frames = []
    with PortalEmprego(ambiente=ambiente) as p:
        for uf in UFS_27:
            df = _df_saldo(p.relatorio(uf, opcional=True))
            if not df.empty:
                df.insert(0, "uf", uf)
                frames.append(df)
    return pd.concat(frames, ignore_index=True) if frames else pd.DataFrame()


def get_relatorio_emprego_estoque_todos_estados(ambiente: str = "producao") -> pd.DataFrame:
    """Estoque de todas as UFs disponíveis, empilhado, com coluna `uf`."""
    frames = []
    with PortalEmprego(ambiente=ambiente) as p:
        for uf in UFS_27:
            df = _df_estoque(p.relatorio(uf, opcional=True))
            if not df.empty:
                df.insert(0, "uf", uf)
                frames.append(df)
    return pd.concat(frames, ignore_index=True) if frames else pd.DataFrame()


# ── Funções de módulo: PORTAL (payload v1.1) ───────────────────────────────────

def get_portal_emprego_estadual(uf: str, ambiente: str = "producao") -> dict:
    """Payload completo do portal (v1.1) de uma UF — como publicado."""
    with PortalEmprego(ambiente=ambiente) as p:
        return p.portal(uf) or {}


def get_portal_emprego_capa_estadual(uf: str, ambiente: str = "producao") -> dict:
    """KPI de capa (`capa_kpi`) — destaque do saldo 12m — de uma UF."""
    with PortalEmprego(ambiente=ambiente) as p:
        return (p.portal(uf) or {}).get("capa_kpi", {})


def get_portal_emprego_kpis_estadual(uf: str, ambiente: str = "producao") -> pd.DataFrame:
    """KPIs do portal de uma UF como DataFrame (id, label, value, display_value, ...)."""
    with PortalEmprego(ambiente=ambiente) as p:
        payload = p.portal(uf) or {}
    return pd.DataFrame(payload.get("kpis") or [])


def get_portal_emprego_charts_estadual(uf: str, ambiente: str = "producao") -> Dict[str, pd.DataFrame]:
    """Séries dos gráficos do portal. dict {chart_id: DataFrame[competencia, valor]}."""
    with PortalEmprego(ambiente=ambiente) as p:
        payload = p.portal(uf) or {}
    out: Dict[str, pd.DataFrame] = {}
    for chart in payload.get("charts") or []:
        series = (chart.get("series") or [{}])[0]
        pts = series.get("points") or []
        out[chart.get("id", f"chart_{len(out)}")] = pd.DataFrame(
            [{"competencia": pt.get("t"), "valor": pt.get("v")} for pt in pts]
        )
    return out


def get_portal_emprego_ranked_lists_estadual(uf: str, ambiente: str = "producao") -> Dict[str, pd.DataFrame]:
    """Rankings do portal (maiores saldos/estoques). dict {list_id: DataFrame[items]}."""
    with PortalEmprego(ambiente=ambiente) as p:
        payload = p.portal(uf) or {}
    return {
        rl.get("id", f"rank_{i}"): pd.DataFrame(rl.get("items") or [])
        for i, rl in enumerate(payload.get("ranked_lists") or [])
    }


def get_portal_emprego_breakdowns_estadual(uf: str, ambiente: str = "producao") -> Dict[str, pd.DataFrame]:
    """Composições do portal (ex.: intensidade tecnológica). dict {id: DataFrame[items]}."""
    with PortalEmprego(ambiente=ambiente) as p:
        payload = p.portal(uf) or {}
    return {
        bd.get("id", f"breakdown_{i}"): pd.DataFrame(bd.get("items") or [])
        for i, bd in enumerate(payload.get("breakdowns") or [])
    }


def get_portal_emprego_kpis_todos_estados(ambiente: str = "producao") -> pd.DataFrame:
    """KPIs de todas as UFs disponíveis, empilhados, com coluna `uf` — comparação entre estados."""
    frames = []
    with PortalEmprego(ambiente=ambiente) as p:
        for uf in UFS_27:
            payload = p.portal(uf, opcional=True)
            if payload and payload.get("kpis"):
                df = pd.DataFrame(payload["kpis"])
                df.insert(0, "uf", uf)
                frames.append(df)
    return pd.concat(frames, ignore_index=True) if frames else pd.DataFrame()


# ── Descoberta ─────────────────────────────────────────────────────────────────

def listar_estados_disponiveis(ambiente: str = "producao", tipo: str = "portal") -> List[str]:
    """Lista as UFs com artefato publicado.

    Args:
        ambiente (str): 'producao' (default) ou 'homologacao'.
        tipo (str): 'portal' (emprego.json) ou 'relatorio' (cache).

    Returns:
        List[str]: siglas das UFs disponíveis.
    """
    disponiveis = []
    with PortalEmprego(ambiente=ambiente) as p:
        for uf in UFS_27:
            payload = (
                p.relatorio(uf, opcional=True)
                if tipo == "relatorio"
                else p.portal(uf, opcional=True)
            )
            if payload is not None:
                disponiveis.append(uf)
    return disponiveis
